// @tesla/react-particles/es/components/Input/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.InputTypes = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _FormError = require('../FormError');

var _FormError2 = _interopRequireDefault(_FormError);

var _input = require('../../scss/_particles/components/_forms/_input.scss');

var _input2 = _interopRequireDefault(_input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var InputTypes = exports.InputTypes = {
  Text: 'text',
  Search: 'search',
  Email: 'email',
  Password: 'password',
  Tel: 'tel',
  Number: 'number'
};

function Input(props) {
  var _inputClasses;

  var inputClasses = (_inputClasses = {}, _defineProperty(_inputClasses, _input2.default['input-rounded'], props.rounded), _defineProperty(_inputClasses, _input2.default['form-error'], props.error), _inputClasses);

  return _react2.default.createElement(
    'div',
    { className: (0, _classnames2.default)(_input2.default['form-item'], props.className), style: props.style },
    _react2.default.createElement(
      'label',
      { htmlFor: props.name, className: _input2.default['form-label'] },
      props.label
    ),
    _react2.default.createElement('input', {
      id: props.name,
      type: props.type,
      name: props.name,
      className: (0, _classnames2.default)(_input2.default['form-text'], inputClasses),
      value: props.value,
      onChange: props.onChange,
      onKeyPress: props.onKeyPress,
      placeholder: props.placeholder,
      disabled: props.disabled,
      defaultValue: props.defaultValue
    }),
    props.error ? _react2.default.createElement(_FormError2.default, { message: props.error }) : null
  );
}

Input.propTypes = {
  type: _propTypes2.default.oneOf(Array.from(Object.keys(InputTypes), function (v) {
    return InputTypes[v];
  })),
  name: _propTypes2.default.string.isRequired,
  label: _propTypes2.default.string,
  value: _propTypes2.default.string,
  defaultValue: _propTypes2.default.string,
  error: _propTypes2.default.string,
  placeholder: _propTypes2.default.string,
  style: _propTypes2.default.shape(),
  onChange: _propTypes2.default.func,
  onKeyPress: _propTypes2.default.func,
  rounded: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  className: _propTypes2.default.string
};

Input.defaultProps = {
  type: InputTypes.Text,
  rounded: false,
  disabled: false,
  label: null,
  value: '',
  defaultValue: undefined,
  error: null,
  placeholder: '',
  style: {},
  onChange: function onChange() {},
  onKeyPress: function onKeyPress() {},
  className: ''
};

exports.default = Input;
