// @tesla/react-particles/es/components/Textarea/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _FormError = require('../FormError');

var _FormError2 = _interopRequireDefault(_FormError);

var _input = require('../../scss/_particles/components/_forms/_input.scss');

var _input2 = _interopRequireDefault(_input);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

function Textarea(props) {
  return _react2.default.createElement(
    'div',
    { className: (0, _classnames2.default)(_input2.default['form-item'], props.className), style: props.style },
    _react2.default.createElement(
      'label',
      { htmlFor: props.name, className: 'form-label' },
      props.label,
      _react2.default.createElement('textarea', {
        id: props.name,
        name: props.name,
        className: (0, _classnames2.default)(_input2.default['form-textarea'], _defineProperty({}, _input2.default['form-error'], props.error)),
        value: props.value,
        onChange: props.onChange,
        onKeyPress: props.onKeyPress,
        placeholder: props.placeholder,
        rows: props.rows
      })
    ),
    props.error ? _react2.default.createElement(_FormError2.default, { message: props.error }) : null
  );
}

Textarea.propTypes = {
  name: _propTypes2.default.string.isRequired,
  label: _propTypes2.default.string,
  value: _propTypes2.default.string,
  error: _propTypes2.default.string,
  placeholder: _propTypes2.default.string,
  rows: _propTypes2.default.number,
  style: _propTypes2.default.shape(),
  onChange: _propTypes2.default.func,
  onKeyPress: _propTypes2.default.func,
  className: _propTypes2.default.string
};

Textarea.defaultProps = {
  label: null,
  value: '',
  error: null,
  placeholder: '',
  rows: 4,
  style: {},
  onChange: function onChange() {},
  onKeyPress: function onKeyPress() {},
  className: null
};

exports.default = Textarea;
