// @tesla/react-particles/es/components/Checkbox/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _checkbox = require('../../scss/_particles/components/_forms/_checkbox.scss');

var _checkbox2 = _interopRequireDefault(_checkbox);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Checkbox(props) {
  return _react2.default.createElement(
    'div',
    { className: _checkbox2.default['form-item'], style: props.style },
    _react2.default.createElement(
      'label',
      { htmlFor: props.name, className: (0, _classnames2.default)(_checkbox2.default['form-label'], _checkbox2.default['label-checkbox']) },
      props.children,
      _react2.default.createElement('input', { id: props.name, name: props.name, type: 'checkbox', className: (0, _classnames2.default)(_checkbox2.default['checkbox'], props.className), onChange: props.onChange, checked: props.checked, readOnly: !props.onChange }),
      _react2.default.createElement('i', { className: _checkbox2.default['icon-checkbox'] })
    )
  );
}

Checkbox.propTypes = {
  name: _propTypes2.default.string.isRequired,
  onChange: _propTypes2.default.func,
  className: _propTypes2.default.string,
  style: _propTypes2.default.shape(),
  checked: _propTypes2.default.bool,
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
};

Checkbox.defaultProps = {
  checked: false,
  onChange: function onChange() {},
  className: '',
  style: {},
  children: null
};

exports.default = Checkbox;
