// @tesla/financial-calculator-components/src/apps/SimpleConfigurator/src/components/Configurator/index.js
import React, { Component } from 'react'
import { connect } from 'react-redux'
import classnames from 'classnames'

import {APP_NAME} from 'dictionary'
import OptionGroup from '../OptionGroup'
import {selectOption} from 'actions'

class Configurator extends Component {

    render() {
        const {groups, currentOptions, selectOption} = this.props
        return (
            <div className="configurator--container">
            <h3 className='configurator--title'>Demo Model 3 Configurator</h3>
            {
                groups.map((group, key)=> <OptionGroup 
                    key={key} 
                    group={group} 
                    currentOptions={currentOptions}
                    selectOption={selectOption}
                 />)
            }
            </div>
        )
    }
}


function mapStateToProps (state, ownProps) {
    const { ui, currentOptions } = state[`${APP_NAME}/Configuration`]
    return {
        groups: ui.groups,
        currentOptions
    }
}


function mapDispatchToProps (dispatch) {
    return {
        selectOption: (option)=>{
            dispatch(selectOption({option}))
        }
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(Configurator)
