// @tesla/react-particles/lib/components/Select/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _select = require('../../scss/_particles/components/_forms/_select.scss');

var _select2 = _interopRequireDefault(_select);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function Select(props) {
  var options = props.options.length > 0 ? props.options : [{
    value: null,
    label: props.emptyLabel
  }];
  return _react2.default.createElement(
    'div',
    { className: (0, _classnames2.default)(_select2.default['form-item'], _select2.default['form-type-select']), style: props.style },
    _react2.default.createElement(
      'label',
      { htmlFor: props.name, className: _select2.default['form-label'] },
      props.label
    ),
    _react2.default.createElement(
      'span',
      { className: _select2.default['form-select-overlay'] },
      _react2.default.createElement(
        'select',
        {
          id: props.name,
          name: props.name,
          className: _select2.default['form-select'],
          value: props.selectedValue,
          onChange: props.onChange
        },
        options.map(function (option) {
          return _react2.default.createElement(
            'option',
            { key: option.value, value: option.value },
            option.label
          );
        })
      )
    )
  );
}

Select.propTypes = {
  name: _propTypes2.default.string.isRequired,
  label: _propTypes2.default.string,
  options: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    value: _propTypes2.default.any,
    label: _propTypes2.default.string
  })),
  selectedValue: _propTypes2.default.string,
  style: _propTypes2.default.shape(),
  onChange: _propTypes2.default.func,
  emptyLabel: _propTypes2.default.string
};

Select.defaultProps = {
  options: [],
  emptyLabel: 'Empty',
  label: '',
  selectedValue: '',
  style: {},
  onChange: function onChange() {}
};

exports.default = Select;
