// @web/tesla-lang/api/rest.js
/**
* Rest endpoints for localization 
* 
* TODO: Build out api to perform currency and number formatting
* TODO: Brainstorm on improving localization workflow
*/
module.exports = function(){
/*
    return {

    	models: {

    		RequestItem: {
    			type: "object",
   			    properties: {
                   type: {
                   	    type: "string",
                   	    description: "The type of format to apply to the value",
                   },
                   value: {
                   	   description: "the value to format"
                   }
    			}
    		},

    		Request: {
    			type: "object",
    			additionalProperties:{
            $ref: "RequestItem"
    			}
    		}
    	},

    	spec: {
        path: "/lang/format/:type",
        "decription": "Formats object values",
        method: "GET",
        parameters: [
           swg.paramTypes.body('Request', "JSON object with ", "Request")
        ]
    	},

    	action: function(req, res){
             var body = req.body;
             for(var key in body){

             }
    	}
    }
  */
};
