// @web/tesla-lang/lib/momentConfig.js
var Utils = require('./Utils');

/**
 * Locale Configuration object for Moment.js
 */
var momentConfig = {


    'da_DK': {
        'months': ['januar', 'februar', 'marts', 'april', 'maj', 'juni', 'juli', 'august', 'september', 'oktober', 'november', 'december'],
        'monthsShort': ['jan', 'feb', 'mar', 'apr', 'maj', 'jun', 'jul', 'aug', 'sep', 'okt', 'nov', 'dec'],
        'weekdays': ['søndag', 'mandag', 'tirsdag', 'onsdag', 'torsdag', 'fredag', 'lørdag'],
        'weekdaysShort': ['søn', 'man', 'tir', 'ons', 'tor', 'fre', 'lør'],
        longDateFormat : {
            LT : 'HH:mm',
            LTS : 'HH:mm:ss',
            L : 'DD/MM/YYYY',
            LL : 'D. MMMM YYYY',
            LLL : 'D. MMMM YYYY HH:mm',
            LLLL : 'dddd [d.] D. MMMM YYYY [kl.] HH:mm'
        }        
    },

    'de': {
        'months': ['Januar', 'Februar', 'März', 'April', 'Mai', 'Juni', 'Juli', 'August', 'September', 'Oktober', 'November', 'Dezember'],
        'monthsShort': ['Jan.', 'Febr.', 'Mrz.', 'Apr.', 'Mai', 'Jun.', 'Jul.', 'Aug.', 'Sept.', 'Okt.', 'Nov.', 'Dez.'],
        'weekdays': ['Sonntag', 'Montag', 'Dienstag', 'Mittwoch', 'Donnerstag', 'Freitag', 'Samstag'],
        'weekdaysShort': ['So.', 'Mo.', 'Di.', 'Mi.', 'Do.', 'Fr.', 'Sa.'],
        'weekdaysMin': ['So', 'Mo', 'Di', 'Mi', 'Do', 'Fr', 'Sa'],
        'longDateFormat': {
            'LT': 'H:mm U\\hr',
            'L': 'DD/MM/YYYY',
            'LL': 'D. MMMM YYYY',
            'LLL': 'D. MMMM YYYY LT',
            'LLLL': 'dddd, D. MMMM YYYY LT'
        },
        'calendar': {
            'sameDay': '[Heute um] LT',
            'sameElse': 'L',
            'nextDay': '[Morgen um] LT',
            'nextWeek': 'dddd [um] LT',
            'lastDay': '[Gestern um] LT',
            'lastWeek': '[letzten] dddd [um] LT'
        },
        'relativeTime': {
            'future': 'in %s',
            'past': 'vor %s',
            's': 'ein paar Sekunden',
            'm': 'einer Minute',
            'mm': '%d Minuten',
            'h': 'einer Stunde',
            'hh': '%d Stunden',
            'd': 'einem Tag',
            'dd': '%d Tagen',
            'M': 'einem Monat',
            'MM': '%d Monaten',
            'y': 'einem Jahr',
            'yy': '%d Jahren'
        }
    },

    'en': {
        'months': ['January', 'February', 'March', 'April', 'May', 'June', 'July', 'August', 'September', 'October', 'November', 'December'],
        'monthsShort': ['Jan', 'Feb', 'Mar', 'Apr', 'May', 'Jun', 'Jul', 'Aug', 'Sep', 'Oct', 'Nov', 'Dec'],
        'weekdays': ['Sunday', 'Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday'],
        'weekdaysShort': ['Sun', 'Mon', 'Tue', 'Wed', 'Thu', 'Fri', 'Sat'],
        'weekdaysMin': ['Su', 'Mo', 'Tu', 'We', 'Th', 'Fr', 'Sa'],
        longDateFormat : {
            LT : 'h:mm A',
            LTS : 'h:mm:ss A',
            L : 'YYYY-MM-DD',
            LL : 'MMMM D, YYYY',
            LLL : 'MMMM D, YYYY h:mm A',
            LLLL : 'dddd, MMMM D, YYYY h:mm A'
        },
        'calendar': {
            'sameDay': '[Today at] LT',
            'nextDay': '[Tomorrow at] LT',
            'nextWeek': 'dddd [at] LT',
            'lastDay': '[Yesterday at] LT',
            'lastWeek': '[last] dddd [at] LT',
            'sameElse': 'L'
        },
        'relativeTime': {
            'future': 'in %s',
            'past': '%s ago',
            's': 'a few seconds',
            'm': 'a minute',
            'mm': '%d minutes',
            'h': 'an hour',
            'hh': '%d hours',
            'd': 'a day',
            'dd': '%d days',
            'M': 'a month',
            'MM': '%d months',
            'y': 'a year',
            'yy': '%d years'
        }
    },

    'es': {
        'months': ['Enero', 'Febrero', 'Marzo', 'Abril', 'Mayo', 'Junio', 'Julio', 'Agosto', 'Septiembre', 'Octubre', 'Noviembre', 'Diciembre'],
        'monthsShort': ['ene', 'feb', 'mar', 'abr', 'may', 'jun', 'jul', 'ago', 'sep', 'oct', 'nov', 'dic'],
        'weekdays': ['domingo', 'lunes', 'martes', 'miércoles', 'jueves', 'viernes', 'sábado'],
        'weekdaysShort': ['dom', 'lun', 'mar', 'mié', 'jue', 'vie', 'sáb'],
        longDateFormat : {
            LT : 'H:mm',
            LTS : 'H:mm:ss',
            L : 'DD/MM/YYYY',
            LL : 'D [de] MMMM [de] YYYY',
            LLL : 'D [de] MMMM [de] YYYY H:mm',
            LLLL : 'dddd, D [de] MMMM [de] YYYY H:mm'
        }
    },

    'fi': {
        'months': ['Tammikuu', 'Helmikuu', 'Maaliskuu', 'Huhtikuu', 'Toukokuu', 'Kesäkuu', 'Heinäkuu', 'Elokuu', 'Syyskuu', 'Lokakuu', 'Marraskuu', 'Joulukuu'],
        'monthsShort': ['Tammi', 'Helmi', 'Maalis', 'Huhti', 'Touko', 'Kesä', 'Heinä', 'Elo', 'Syys', 'Loka', 'Marras', 'Joulu'],
        'weekdays': ['Sunnuntai', 'Maanantai', 'Tiistai', 'Keskiviikko', 'Torstai', 'Perjantai', 'Lauantai'],
        'weekdaysShort': ['su', 'ma', 'ti', 'ke', 'to', 'pe', 'la'],
        'weekdaysMin': ['su', 'ma', 'ti', 'ke', 'to', 'pe', 'la'],
        longDateFormat : {
            LT : 'HH.mm',
            LTS : 'HH.mm.ss',
            L : 'DD.MM.YYYY',
            LL : 'Do MMMM[ta] YYYY',
            LLL : 'Do MMMM[ta] YYYY, [klo] HH.mm',
            LLLL : 'dddd, Do MMMM[ta] YYYY, [klo] HH.mm',
            l : 'D.M.YYYY',
            ll : 'Do MMM YYYY',
            lll : 'Do MMM YYYY, [klo] HH.mm',
            llll : 'ddd, Do MMM YYYY, [klo] HH.mm'
        }        
    },

    'fr': {
        'months': ['Janvier', 'Février', 'Mars', 'Avril', 'Mai', 'Juin', 'Juillet', 'Aout', 'Septembre', 'Octobre', 'Novembre', 'Décembre'],
        'monthsShort': ['Jan', 'Fev', 'Mar', 'Avr', 'Mai', 'Juin', 'Juil', 'Aou', 'Sep', 'Oct', 'Nov', 'Dec'],
        'weekdays': ['Dimanche', 'Lundi', 'Mardi', 'Mercredi', 'Jeudi', 'Vendredi', 'Samedi'],
        'weekdaysShort': ['Dim', 'Lun', 'Mar', 'Mer', 'Jeu', 'Ven', 'Sam'],
        'longDateFormat': {
            'LT': 'HH:mm',
            'LTS': 'HH:mm:ss',
            'L': 'DD/MM/YYYY',
            'LL': 'D MMMM YYYY',
            'LLL': 'D MMMM YYYY HH:mm',
            'LLLL': 'dddd D MMMM YYYY HH:mm'
        },
        'calendar': {
            'sameDay': '[Aujourd\'hui à] LT',
            'nextDay': '[Demain à] LT',
            'nextWeek': 'dddd [à] LT',
            'lastDay': '[Hier à] LT',
            'lastWeek': 'dddd [dernier à] LT',
            'sameElse': 'L'
        },
        'relativeTime': {
            'future': 'dans %s',
            'past': 'il y a %s',
            's': 'quelques secondes',
            'm': 'une minute',
            'mm': '%d minutes',
            'h': 'une heure',
            'hh': '%d heures',
            'd': 'un jour',
            'dd': '%d jours',
            'M': 'un mois',
            'MM': '%d mois',
            'y': 'un an',
            'yy': '%d ans'
        },

        'ordinalParse': /\d{1,2}(er|)/,

        'week': {
            'dow': 1, // Monday is the first day of the week.
            'doy': 4 // The week that contains Jan 4th is the first week of the year.
        }

    },

    'it': {
        'months': ['Gennaio', 'Febbraio', 'Marzo', 'Aprile', 'Maggio', 'Giugno', 'Luglio', 'Agosto', 'Settembre', 'Ottobre', 'Novembre', 'Dicembre'],
        'monthsShort': ['Gen', 'Feb', 'Mar', 'Apr', 'Mag', 'Giu', 'Lug', 'Ago', 'Set', 'Ott', 'Nov', 'Dic'],
        'weekdays': ['Domenica', 'Lunedì', 'Martedì', 'Mercoledì', 'Giovedì', 'Venerdì', 'Sabato'],
        'weekdaysShort': ['Dom', 'Lun', 'Mar', 'Mer', 'Gio', 'Ven', 'Sab'],
        'weekdaysMin': ['D', 'L', 'Ma', 'Me', 'G', 'V', 'S'],
        'longDateFormat': {
            'LT': 'HH:mm',
            'L': 'DD/MM/YYYY',
            'LL': 'D MMMM YYYY',
            'LLL': 'D MMMM YYYY LT',
            'LLLL': 'dddd, D MMMM YYYY LT'
        },
        'calendar': {
            'sameDay': '[Oggi alle] LT',
            'nextDay': '[Domani alle] LT',
            'nextWeek': 'dddd [alle] LT',
            'lastDay': '[Ieri alle] LT',
            'lastWeek': '[lo scorso] dddd [alle] LT',
            'sameElse': 'L'
        },
        'relativeTime': {
            'future': 'in %s',
            'past': '%s fa',
            's': 'secondi',
            'm': 'un minuto',
            'mm': '%d minuti',
            'h': 'un\'ora',
            'hh': '%d ore',
            'd': 'un giorno',
            'dd': '%d giorni',
            'M': 'un mese',
            'MM': '%d mesi',
            'y': 'un anno',
            'yy': '%d anni'
        }
    },

    'ja': {
        months: ['1月', '2月', '3月', '4月', '5月', '6月', '7月', '8月', '9月', '10月', '11月', '12月'],
        monthsShort: ['1月', '2月', '3月', '4月', '5月', '6月', '7月', '8月', '9月', '10月', '11月', '12月'],
        weekdays: ['日曜日', '月曜日', '火曜日', '水曜日', '木曜日', '金曜日', '土曜日'],
        weekdaysShort: ['日', '月', '火', '水', '木', '金', '土'],
        weekdaysMin: ['日', '月', '火', '水', '木', '金', '土']
    },

    'ko': {
        months: ['1월', '2월', '3월', '4월', '5월', '6월', '7월', '8월', '9월', '10월', '11월', '12월'],
        monthsShort: ['1월', '2월', '3월', '4월', '5월', '6월', '7월', '8월', '9월', '10월', '11월', '12월'],
        weekdays: ['일요일', '월요일', '화요일', '수요일', '목요일', '금요일', '토요일'],
        weekdaysShort: ['일', '월', '화', '수', '목', '금', '토'],
        weekdaysMin: ['일', '월', '화', '수', '목', '금', '토'],
        longDateFormat : {
            LT : 'A h:mm',
            LTS : 'A h:mm:ss',
            L : 'YYYY.MM.DD',
            LL : 'YYYY년 MMMM D일',
            LLL : 'YYYY년 MMMM D일 A h:mm',
            LLLL : 'YYYY년 MMMM D일 dddd A h:mm',
            l : 'YYYY.MM.DD',
            ll : 'YYYY년 MMMM D일',
            lll : 'YYYY년 MMMM D일 A h:mm',
            llll : 'YYYY년 MMMM D일 dddd A h:mm'
        }        
    },

    'nl': {
        'months': ['januari', 'februari', 'maart', 'april', 'mei', 'juni', 'juli', 'augustus', 'september', 'oktober', 'november', 'december'],
        'weekdays': ['zondag', 'maandag', 'dinsdag', 'woensdag', 'donderdag', 'vrijdag', 'zaterdag'],
        'weekdaysShort': ['zo.', 'ma.', 'di.', 'wo.', 'do.', 'vr.', 'za.'],
        'weekdaysMin': ['Zo', 'Ma', 'Di', 'Wo', 'Do', 'Vr', 'Za'],
        'longDateFormat': {
            'LT': 'HH:mm',
            'L': 'DD/MM/YYYY',
            'LL': 'D MMMM YYYY',
            'LLL': 'D MMMM YYYY LT',
            'LLLL': 'dddd D MMMM YYYY LT'
        },
        'calendar': {
            'sameDay': '[Vandaag om] LT',
            'nextDay': '[Morgen om] LT',
            'nextWeek': 'dddd [om] LT',
            'lastDay': '[Gisteren om] LT',
            'lastWeek': '[afgelopen] dddd [om] LT',
            'sameElse': 'L'
        },
        'relativeTime': {
            'future': 'over %s',
            'past': '%s geleden',
            's': 'een paar seconden',
            'm': 'één minuut',
            'mm': '%d minuten',
            'h': 'één uur',
            'hh': '%d uur',
            'd': 'één dag',
            'dd': '%d dagen',
            'M': 'één maand',
            'MM': '%d maanden',
            'y': 'één jaar',
            'yy': '%d jaar'
        }
    },

    'no': {
        'months': ['januar', 'februar', 'mars', 'april', 'mai', 'juni', 'juli', 'august', 'september', 'oktober', 'november', 'desember'],
        'monthsShort': ['jan', 'feb', 'mar', 'apr', 'mai', 'jun', 'jul', 'aug', 'sep', 'okt', 'nov', 'des'],
        'weekdays': ['søndag', 'mandag', 'tirsdag', 'onsdag', 'torsdag', 'fredag', 'lørdag'],
        'weekdaysShort': ['søn', 'man', 'tir', 'ons', 'tor', 'fre', 'lør'],
        'weekdaysMin': ['sø', 'ma', 'ti', 'on', 'to', 'fr', 'lø'],
        'longDateFormat': {
            'LT': 'HH:mm',
            'L': 'DD/MM/YYYY',
            'LL': 'D MMMM YYYY',
            'LLL': 'D MMMM YYYY LT',
            'LLLL': 'dddd D MMMM YYYY LT'
        },
        'calendar': {
            'sameDay': '[I dag klokken] LT',
            'nextDay': '[I morgen klokken] LT',
            'nextWeek': 'dddd [klokken] LT',
            'lastDay': '[I går klokken] LT',
            'lastWeek': '[Forrige] dddd [klokken] LT',
            'sameElse': 'L:'
        },
        'relativeTime': {
            'future': 'om %s',
            'past': 'for %s siden',
            's': 'noen sekunder',
            'm': 'ett minutt',
            'mm': '%d minutter',
            'h': 'en time',
            'hh': '%d timer',
            'd': 'en dag',
            'dd': '%d dager',
            'M': 'en måned',
            'MM': '%d måneder',
            'y': 'ett år',
            'yy': '%d år'
        }
    },

    'pt': {
        months: ['Janeiro', 'Fevereiro', 'Março', 'Abril', 'Maio', 'Junho', 'Julho', 'Agosto', 'Setembro', 'Outubro', 'Novembro', 'Dezembro'],
        monthsShort: ['Jan', 'Fev', 'Mar', 'Abr', 'Mai', 'Jun', 'Jul', 'Ago', 'Set', 'Out', 'Nov', 'Dez'],
        weekdays: ['Domingo', 'Segunda-Feira', 'Terça-Feira', 'Quarta-Feira', 'Quinta-Feira', 'Sexta-Feira', 'Sábado'],
        weekdaysShort: ['Dom', 'Seg', 'Ter', 'Qua', 'Qui', 'Sex', 'Sáb'],
        weekdaysMin: ['Do', '2ª', '3ª', '4ª', '5ª', '6ª', 'Sá'],
        longDateFormat : {
            LT : 'HH:mm',
            LTS : 'HH:mm:ss',
            L : 'DD/MM/YYYY',
            LL : 'D [de] MMMM [de] YYYY',
            LLL : 'D [de] MMMM [de] YYYY HH:mm',
            LLLL : 'dddd, D [de] MMMM [de] YYYY HH:mm'
        }        
    },

    'sv': {
        months: ['januari', 'februari', 'mars', 'april', 'maj', 'juni', 'juli', 'augusti', 'september', 'oktober', 'november', 'december'],
        monthsShort: ['jan', 'feb', 'mar', 'apr', 'maj', 'jun', 'jul', 'aug', 'sep', 'okt', 'nov', 'dec'],
        weekdays: ['söndag', 'måndag', 'tisdag', 'onsdag', 'torsdag', 'fredag', 'lördag'],
        weekdaysShort: ['sön', 'mån', 'tis', 'ons', 'tor', 'fre', 'lör'],
        weekdaysMin: ['sö', 'må', 'ti', 'on', 'to', 'fr', 'lö'],
        longDateFormat : {
            LT : 'HH:mm',
            LTS : 'HH:mm:ss',
            L : 'YYYY-MM-DD',
            LL : 'D MMMM YYYY',
            LLL : 'D MMMM YYYY [kl.] HH:mm',
            LLLL : 'dddd D MMMM YYYY [kl.] HH:mm',
            lll : 'D MMM YYYY HH:mm',
            llll : 'ddd D MMM YYYY HH:mm'
        }     
    },

    'zh': {
        months: ['一月', '二月', '三月', '四月', '五月', '六月', '七月', '八月', '九月', '十月', '十一月', '十二月'],
        monthsShort: ['1月', '2月', '3月', '4月', '5月', '6月', '7月', '8月', '9月', '10月', '11月', '12月'],
        weekdays: ['星期日', '星期一', '星期二', '星期三', '星期四', '星期五', '星期六'],
        weekdaysShort: ['周日', '周一', '周二', '周三', '周四', '周五', '周六'],
        weekdaysMin: ['日', '一', '二', '三', '四', '五', '六'],
        longDateFormat: {
            LT: 'Ah点mm',
            LTS: 'Ah点m分s秒',
            L: 'YYYY-MM-DD',
            LL: 'YYYY年MMMD日',
            LLL: 'YYYY年MMMD日LT',
            LLLL: 'YYYY年MMMD日ddddLT',
            l: 'YYYY-MM-DD',
            ll: 'YYYY年MMMD日',
            lll: 'YYYY年MMMD日LT',
            llll: 'YYYY年MMMD日ddddLT'
        }
    }
};

/**
 * Returns configuration for moment.js
 * @param  {String} locale [description]
 * @return {Object} 
 */
module.exports.getLocaleSettings = function(locale) {

    // check if there is a config that exactly matches the locale e.g., 'en_US'
    var cfg = momentConfig[locale];

    if (!cfg) {

        // if particular language + market locale does not exist, default to locale that matches language
        var loc = Utils.parseLocale(locale);

        if (loc) {

            // check for config that matches language code e.g.: 'en', 'de', etc...
            cfg = momentConfig[loc.lang];

            // look for config with same language code as locale key
            if (!cfg) {
                var keys = Object.keys(momentConfig);
                for (var i = 0; i < keys.length; i++) {
                    var key = keys[i];
                    if (key.indexOf(loc.lang) === 0) {
                        cfg = momentConfig[key];
                        break;
                    }
                }
            }

            // Still no matching config? Default to 'en'
            if (!cfg) {
                cfg = momentConfig.en;
                console.error('Date format for ', locale, ' not found. Defaulting to en');
            }
        }
    }

    return cfg;
};

module.exports.configs = momentConfig;
