// @web/tesla-lang/lib/UnitConverter.js
/**
 * Basic unit conversions and utility methods
 *
 * Note: 
 *    If we find ourselves having to do more than these basic conversions, 
 *    we should consider using one of the excellent libs such as 
 *     js-quantities, convert-units, etc...
 *   
 * 
 * @author Frank Young 03/2016
 */
module.exports = {
	
    /**
     * Returns scaled SI Units for kW, kWh
     * given a value and unit string (1,000kW => 1MW)
     *
     * Example:
     *
     *    NumberFormat.getScaledUnits(1837, 'kW'); // returns 1.8 MW
     *
     * @param {Number} val value
     * @param {String} units unit type
     * @param {Number} rounding how many places to round to
     * @return {Number}
     */
    getScaledUnits: function(val, units, rounding){

        rounding = rounding || 100;

        var needsScaling = val >= 1000;
        var scaledValue = needsScaling ? Math.round( (val / 1000) * rounding) / rounding : val;

        // kilo, mega, giga
        var siUnitsPrefix = ['k', 'M', 'G'];
        var siUnitsSuffix = units.substring(1);
        var siUnitsPrefixIndex = siUnitsPrefix.indexOf(units.substring(0,1));

        if(needsScaling && siUnitsPrefixIndex < siUnitsPrefix.length){
            units = siUnitsPrefix[siUnitsPrefixIndex + 1] + siUnitsSuffix;
        }

        // if value still needs scaling
        if(scaledValue >= 1000){
            return this.getScaledUnits(scaledValue, units);
        }

        return {
            units: units,
            value: scaledValue
        };
    },
     
     /**
      * Convert Fahrenheit to Celcius
      * @param  {Number} tmp
      * @return {Number}
      */
     fahrenheitToCelcius: function(tmp){
         return Math.round( ((tmp - 32) * .5556) * 100) / 100;
     },

     /**
      * Convert Celcius to Fahrenheit
      * @param  {Number} tmp
      * @return {Number}
      */
     celciusToFahrenheit: function(tmp){
         return Math.round( ((tmp * 1.8) + 32) * 100 / 100);
     },

     /**
      * Convert miles to kilometers
      * @param  {Number} miles
      * @return {Number}
      */
     milesToKilometers: function(miles){
         return Math.round( (miles * 1.60934) * 100) / 100;
     },

     /**
      * Convert kilometers to miles
      * @param  {Number} km
      * @return {Number}
      */
     kilometersToMiles: function(km){
        return Math.round( (km / 1.60934) * 100) / 100;
     },

     /**
      * Convert lbs to kilograms 
      * @param  {Number} lbs
      * @return {Number} lbs converted to kilograms 
      */
     poundsToKilograms: function(lbs){
          return Math.round( (lbs * 0.453592) * 100) / 100;
     },

     /**
      * Convert kilograms to lbs
      * @param  {Number} kg
      * @return {Number} kg converted to pounds 
      */
     kilogramsToPounds: function(kg){
          return Math.round( (kg * 2.20462) * 100) / 100;
     }     

};
