// @web/tesla-lang/lib/Utils.js
function validateZip(market, zipCode){

    zipCode = zipCode || '';

    var map = {
        US: /^([0-9]{5})(-[0-9]{4})?$/
    };

    if(map[market]){
        var match = zipCode.match(map[market]);
        return match ? true : false;
    }
    else {
        console.warn('validateZip requires market');
    }

    return false;
}

/**
 * Returns object with lang and country properties from locale string
 * @param {String} locale (langCode_CountryCode: [en_US, de_DE, it_IT, etc...])
 */
function parseLocale(locale, delimiter) {

    delimiter = delimiter || '_';

    if(!locale){
        console.error('parseLocale: missing locale param');
        return null;
    }

    var parsed = locale.split(delimiter)
       ,lang
       ,country;

    if (parsed.length === 2) {
        country = parsed[1].toUpperCase();
    }
    else {      
        country = locale.toUpperCase();
    }

    lang = parsed[0].toLowerCase();

    return {
        locale: lang + '_' + country,
        lang: lang,
        country: country
    };
}


module.exports = {
    validateZip: validateZip,
    parseLocale: parseLocale
};
