// @tesla/react-particles/lib/components/Table/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TableCell = exports.TableRow = exports.TableBody = exports.TableHeadCell = exports.TableHead = exports.TableTypes = undefined;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _tables = require('../../scss/_particles/components/_tables.scss');

var _tables2 = _interopRequireDefault(_tables);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var TableTypes = exports.TableTypes = {
  Data: 'data',
  Forum: 'forum'
};

function Table(props) {
  var columns = !props.children ? props.columns : null;
  var data = !props.children ? props.data : null;
  var headerComponent = tableHeaderComponent(columns);
  var bodyComponent = tableBodyComponent(columns, data);

  var tableClassNames = {
    table: _tables2.default['table']
  };
  if (props.type) {
    tableClassNames[_tables2.default['table-' + props.type]] = true;
  }
  return _react2.default.createElement(
    'table',
    { className: (0, _classnames2.default)(props.className, tableClassNames), style: props.style },
    props.children,
    headerComponent,
    bodyComponent
  );

  function tableHeaderComponent(columns) {
    if (!columns) {
      return;
    }

    var tableHeaderComponents = columns.map(tableHeaderCellComponent);
    return _react2.default.createElement(
      TableHead,
      null,
      _react2.default.createElement(
        TableRow,
        null,
        tableHeaderComponents
      )
    );
  }

  function tableHeaderCellComponent(columnHeader, key) {
    var value = void 0;

    if (typeof columnHeader.header === 'string') {
      value = columnHeader.header;
    }

    if (typeof columnHeader.header === 'function') {
      value = columnHeader.header();
    }

    return _react2.default.createElement(
      TableHeadCell,
      { key: key },
      value
    );
  }

  function tableBodyComponent(columns, data) {
    if (!columns || !data) {
      return;
    }

    var tableRowComponents = data.map(function (row, idx) {
      return tableRowComponent(columns, row, idx);
    });
    return _react2.default.createElement(
      TableBody,
      null,
      tableRowComponents
    );
  }

  function tableRowComponent(columns, row, key) {
    var cellComponents = columns.map(function (column, idx) {
      return tableCellComponent(column, row, idx);
    });
    return _react2.default.createElement(
      TableRow,
      { key: key },
      cellComponents
    );
  }

  function tableCellComponent(column, row, idx) {
    var value = void 0;

    if (typeof column.accessor === 'string') {
      value = column.accessor.split('.').reduce(function (a, b) {
        return a[b];
      }, row);
    }

    if (typeof column.accessor === 'function') {
      value = column.accessor(row, idx);
    }

    return _react2.default.createElement(
      TableCell,
      { key: idx, style: column.style },
      value
    );
  }
}

Table.propTypes = {
  columns: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    header: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]),
    accessor: _propTypes2.default.oneOfType([_propTypes2.default.string, _propTypes2.default.func]),
    style: _propTypes2.default.any
  })),
  data: _propTypes2.default.arrayOf(_propTypes2.default.any),
  type: _propTypes2.default.oneOf(Array.from(Object.keys(TableTypes), function (v) {
    return TableTypes[v];
  })),
  style: _propTypes2.default.any
};

function TableHead(props) {
  return _react2.default.createElement(
    'thead',
    { className: _tables2.default['table-head'] },
    props.children
  );
}

TableHead.propTypes = {
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
};

TableHead.defaultProps = {
  children: null
};

function TableBody(props) {
  return _react2.default.createElement(
    'tbody',
    { className: _tables2.default['table-body'] },
    props.children
  );
}

TableBody.propTypes = {
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
};

TableBody.defaultProps = {
  children: null
};

function TableRow(props) {
  return _react2.default.createElement(
    'tr',
    { className: _tables2.default['table-row'] },
    props.children
  );
}

TableRow.propTypes = {
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
};

TableRow.defaultProps = {
  children: null
};

function TableHeadCell(props) {
  return _react2.default.createElement(
    'th',
    null,
    props.children
  );
}

TableHeadCell.propTypes = {
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
};

TableHeadCell.defaultProps = {
  children: null
};

function TableCell(props) {
  return _react2.default.createElement(
    'td',
    { style: props.style },
    props.children
  );
}

TableCell.propTypes = {
  style: _propTypes2.default.shape(),
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
};

TableCell.defaultProps = {
  style: {},
  children: null
};

exports.default = Table;
exports.TableHead = TableHead;
exports.TableHeadCell = TableHeadCell;
exports.TableBody = TableBody;
exports.TableRow = TableRow;
exports.TableCell = TableCell;
