// @tesla/react-particles/es/components/Button/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.FormTypes = exports.OutlineStyle = exports.ButtonSizes = exports.ButtonTypes = undefined;
exports.ButtonGroup = ButtonGroup;

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _buttons = require('../../scss/_particles/components/_buttons.scss');

var _buttons2 = _interopRequireDefault(_buttons);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var ButtonTypes = exports.ButtonTypes = {
  Primary: 'primary',
  Secondary: 'secondary',
  Tertiary: 'tertiary',
  Transactional: 'transactional'
};

var ButtonSizes = exports.ButtonSizes = {
  Large: 'large',
  Medium: 'medium',
  Small: 'small',
  Tiny: 'tiny'
};

var OutlineStyle = exports.OutlineStyle = {
  Black: 'black',
  Alternate: 'alternate',
  White: 'white',
  Grey: 'grey'
};

var FormTypes = exports.FormTypes = {
  Button: 'button',
  Submit: 'submit',
  Reset: 'reset',
  Menu: 'menu'
};

function ButtonGroup(props) {
  return _react2.default.createElement(
    'div',
    { className: _buttons2.default['btn-group'], style: props.style },
    props.children
  );
}

ButtonGroup.propTypes = {
  style: _propTypes2.default.shape(),
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
};

ButtonGroup.defaultProps = {
  style: {},
  children: null
};

function Button(props) {
  var _buttonClasses;

  var buttonClasses = (_buttonClasses = {}, _defineProperty(_buttonClasses, _buttons2.default['btn-' + props.type], true), _defineProperty(_buttonClasses, _buttons2.default['btn-' + props.size], !!props.size), _defineProperty(_buttonClasses, _buttons2.default['btn-full'], props.fullWidth), _defineProperty(_buttonClasses, _buttons2.default['btn-outline'], props.outline), _defineProperty(_buttonClasses, _buttons2.default['outline-' + props.outlineStyle], props.outline && !!props.outlineStyle), _defineProperty(_buttonClasses, _buttons2.default['btn-transparent'], props.transparent), _defineProperty(_buttonClasses, _buttons2.default['transparent-dark'], props.transparent && props.transparentDark), _buttonClasses);

  return _react2.default.createElement(
    'button',
    {
      type: props.formType,
      className: (0, _classnames2.default)(buttonClasses),
      onClick: props.onClick,
      style: props.style,
      disabled: props.disabled
    },
    props.children
  );
}

Button.propTypes = {
  type: _propTypes2.default.oneOf(Array.from(Object.keys(ButtonTypes), function (v) {
    return ButtonTypes[v];
  })),
  size: _propTypes2.default.oneOf(Array.from(Object.keys(ButtonSizes), function (v) {
    return ButtonSizes[v];
  })),
  fullWidth: _propTypes2.default.bool,
  outline: _propTypes2.default.bool,
  disabled: _propTypes2.default.bool,
  transparent: _propTypes2.default.bool,
  transparentDark: _propTypes2.default.bool,
  outlineStyle: _propTypes2.default.oneOf(Array.from(Object.keys(OutlineStyle), function (v) {
    return OutlineStyle[v];
  })),
  onClick: _propTypes2.default.func,
  style: _propTypes2.default.shape(),
  formType: _propTypes2.default.oneOf(Array.from(Object.keys(FormTypes), function (v) {
    return FormTypes[v];
  })),
  children: _propTypes2.default.oneOfType([_propTypes2.default.arrayOf(_propTypes2.default.node), _propTypes2.default.node])
};

Button.defaultProps = {
  type: ButtonTypes.Primary,
  size: null,
  fullWidth: false,
  outline: false,
  transparent: false,
  transparentDark: false,
  disabled: false,
  outlineStyle: null,
  onClick: function onClick() {},
  style: {},
  formType: FormTypes.Button,
  children: null
};

exports.default = Button;
